daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/ImmArray.scala

```diff
@@ -202,23 +201,16 @@ final class ImmArray[+A] private (
     * since to snoc an ImmArray we must copy it.
     */
   def slowSnoc[B >: A](el: B): ImmArray[B] = {
-    val newArray: mutable.ArraySeq[B] = new mutable.ArraySeq(length + 1)
-    for (i <- indices) {
-      newArray(i) = uncheckedGet(i)
-    }
-    newArray(length) = el
-    ImmArray.unsafeFromArraySeq(newArray)
+    val newLen = length + 1
+    val newArray = new Array[AnyRef](newLen)
+    copyToAnyRefArray(newArray, 0, length)
+    newArray(length) = el.asInstanceOf[AnyRef]
+    new ImmArray[B](0, newLen, newArray)
   }
 
   /** O(min(n, m)) */
-  def zip[B](that: ImmArray[B]): ImmArray[(A, B)] = {
-    val newLen = Math.min(length, that.length)
-    val newArray: mutable.ArraySeq[(A, B)] = new mutable.ArraySeq(newLen)
-    for (i <- 0 until newLen) {
-      newArray(i) = (uncheckedGet(i), that.uncheckedGet(i))
-    }
-    ImmArray.unsafeFromArraySeq(newArray)
-  }
+  def zip[B](that: ImmArray[B]): ImmArray[(A, B)] =
+    ImmArray.fromIterator(this.iterator zip that.iterator, this.length min that.length)
```

Comment:
Wouldn't a hand written loop similar to what we had before but using a `while`-loop be faster? It avoids all the indirections through the iterator.